File: cpp/base/src/main/vsdk/toolkit/media/RGBImageUncompressed.cpp
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/media/RGBPixel.h"
#include "vsdk/toolkit/media/RGBAImageUncompressed.h"

RGBImageUncompressed::RGBImageUncompressed() :
    data(nullptr), xSize(0), ySize(0), rowStride(0) {
}

RGBImageUncompressed::~RGBImageUncompressed() {
    if (data != nullptr) {
        delete[] data;
        data = nullptr;
    }
}

void RGBImageUncompressed::detach() {
    if (data != nullptr) {
        delete[] data;
        data = nullptr;
    }
}

int RGBImageUncompressed::getSizeInBytes() const {
    return xSize * ySize * 3 + 2 * INT_SIZE_IN_BYTES + POINTER_SIZE_IN_BYTES;
}

bool RGBImageUncompressed::init(int width, int height) {
    try {
        if (data != nullptr) {
            delete[] data;
        }
        data = new char[width * height * BYTES_PER_PIXEL];
        for (int i = 0; i < width * height * BYTES_PER_PIXEL; i++) {
            data[i] = 0;
        }
    } catch (...) {
        data = nullptr;
        return false;
    }
    xSize = width;
    ySize = height;
    rowStride = width * BYTES_PER_PIXEL;
    return true;
}

bool RGBImageUncompressed::initNoFill(int width, int height) {
    if (data != nullptr && width == xSize && height == ySize) {
        return true;
    }

    try {
        if (data != nullptr) {
            delete[] data;
        }
        data = new char[width * height * BYTES_PER_PIXEL];
    } catch (...) {
        data = nullptr;
        return false;
    }
    xSize = width;
    ySize = height;
    rowStride = width * BYTES_PER_PIXEL;
    return true;
}

int RGBImageUncompressed::pixelBaseIndex(int x, int y) const {
    return (ySize - 1 - y) * rowStride + x * BYTES_PER_PIXEL;
}

void RGBImageUncompressed::putPixel(int x, int y, char r, char g, char b) {
    int index = pixelBaseIndex(x, y);
    data[index] = r;
    data[index + 1] = g;
    data[index + 2] = b;
}

void RGBImageUncompressed::putPixel(int x, int y, const RGBPixel* p) {
    int index = pixelBaseIndex(x, y);
    data[index] = p->r;
    data[index + 1] = p->g;
    data[index + 2] = p->b;
}

void RGBImageUncompressed::putPixelRgb(int x, int y, RGBPixel* p) {
    int index = pixelBaseIndex(x, y);
    data[index] = p->r;
    data[index + 1] = p->g;
    data[index + 2] = p->b;
}

RGBPixel* RGBImageUncompressed::getPixel(int x, int y) const {
    RGBPixel* p = new RGBPixel();
    int index = pixelBaseIndex(x, y);
    p->r = data[index];
    p->g = data[index + 1];
    p->b = data[index + 2];
    return p;
}

RGBPixel* RGBImageUncompressed::getPixelRgb(int x, int y) const {
    RGBPixel* p = new RGBPixel();
    int index = pixelBaseIndex(x, y);
    p->r = data[index];
    p->g = data[index + 1];
    p->b = data[index + 2];
    return p;
}

void RGBImageUncompressed::getPixelRgb(int x, int y, RGBPixel* p) const {
    int index = pixelBaseIndex(x, y);
    p->r = data[index];
    p->g = data[index + 1];
    p->b = data[index + 2];
}

int RGBImageUncompressed::getXSize() const {
    return xSize;
}

int RGBImageUncompressed::getYSize() const {
    return ySize;
}

char* RGBImageUncompressed::getRawImage() const {
    if (data == nullptr) {
        return nullptr;
    }
    char* copy = new char[xSize * ySize * BYTES_PER_PIXEL];
    for (int i = 0; i < xSize * ySize * BYTES_PER_PIXEL; i++) {
        copy[i] = data[i];
    }
    return copy;
}

void RGBImageUncompressed::setRawImage(int width, int height, char* imageData) {
    xSize = width;
    ySize = height;
    rowStride = width * BYTES_PER_PIXEL;

    if (data != nullptr) {
        delete[] data;
    }
    if (imageData != nullptr) {
        int size = width * height * BYTES_PER_PIXEL;
        data = new char[size];
        for (int i = 0; i < size; i++) {
            data[i] = imageData[i];
        }
    } else {
        data = nullptr;
    }
}

RGBImageUncompressed* RGBImageUncompressed::clone() const {
    RGBImageUncompressed* copy = new RGBImageUncompressed();
    int xxSize = getXSize();
    int yySize = getYSize();
    copy->init(xxSize, yySize);

    for (int x = 0; x < xxSize; x++) {
        for (int y = 0; y < yySize; y++) {
            RGBPixel* pixel = getPixel(x, y);
            copy->putPixel(x, y, pixel);
            delete pixel;
        }
    }
    return copy;
}

RGBAImageUncompressed* RGBImageUncompressed::cloneToRgba() const {
    RGBAImageUncompressed* copy = new RGBAImageUncompressed();
    int xxSize = getXSize();
    int yySize = getYSize();
    copy->init(xxSize, yySize);

    for (int x = 0; x < xxSize; x++) {
        for (int y = 0; y < yySize; y++) {
            RGBPixel* source = getPixel(x, y);
            copy->putPixelA(x, y, source->r, source->g, source->b, (char)255);
            delete source;
        }
    }
    return copy;
}

void RGBImageUncompressed::dispose() {
}

File: typescript/base/src/main/vsdk/toolkit/media/RGBImageUncompressed.ts
import { Image } from "./Image.js";
import { RGBPixel } from "./RGBPixel.js";
import { byte } from "./_pixel.js";
import { RGBAImageUncompressed } from "./RGBAImageUncompressed.js";
import { RGBAPixel } from "./RGBAPixel.js";
export class RGBImageUncompressed extends Image {
    private data: Uint8Array | null = null;
    private xSize = 0;
    private ySize = 0;
    private rowStride = 0;
    public dettach(): void {
        this.data = null;
    }
    public override getSizeInBytes(): number {
        return this.xSize * this.ySize * 3 + 16;
    }
    public init(w: number, h: number): boolean {
        return this.allocate(w, h);
    }
    public initNoFill(w: number, h: number): boolean {
        return this.data !== null && w === this.xSize && h === this.ySize ? true : this.allocate(w, h);
    }
    private allocate(w: number, h: number): boolean {
        try {
            this.data = new Uint8Array(w * h * 3);
            this.xSize = w;
            this.ySize = h;
            this.rowStride = w * 3;
            return true;
        } catch {
            this.data = null;
            return false;
        }
    }
    private at(x: number, y: number): number {
        return (this.ySize - 1 - y) * this.rowStride + x * 3;
    }
    public putPixel(x: number, y: number, r: number | RGBPixel, g?: number, b?: number): void {
        const i = this.at(x, y),
            d = this.data!;
        if (r instanceof RGBPixel) {
            d[i] = r.r;
            d[i + 1] = r.g;
            d[i + 2] = r.b;
        } else {
            // Java only declares the `byte` overload here, which stores the
            // signed byte as given; the unsigned backing array wraps it back.
            d[i] = r;
            d[i + 1] = g!;
            d[i + 2] = b!;
        }
    }
    public override putPixelRgb(x: number, y: number, p: RGBPixel): void {
        this.putPixel(x, y, p);
    }
    public getPixel(x: number, y: number): RGBPixel {
        const i = this.at(x, y),
            d = this.data!,
            p = new RGBPixel();
        p.r = byte(d[i]!);
        p.g = byte(d[i + 1]!);
        p.b = byte(d[i + 2]!);
        return p;
    }
    public override getPixelRgb(x: number, y: number): RGBPixel;
    public override getPixelRgb(x: number, y: number, p: RGBPixel): void;
    public override getPixelRgb(x: number, y: number, p?: RGBPixel): RGBPixel | void {
        const v = this.getPixel(x, y);
        if (p) {
            p.r = v.r;
            p.g = v.g;
            p.b = v.b;
            return;
        }
        return v;
    }
    public getXSize(): number {
        return this.xSize;
    }
    public getYSize(): number {
        return this.ySize;
    }
    public getRawImage(): Int8Array {
        return new Int8Array(this.data!.buffer, this.data!.byteOffset, this.data!.length);
    }
    public getRawImageDirectBuffer(): Uint8Array {
        return this.data!;
    }
    public setRawImage(w: number, h: number, data: Int8Array | Uint8Array): void {
        this.xSize = w;
        this.ySize = h;
        this.rowStride = w * 3;
        this.data = new Uint8Array(data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength));
    }
    public override clone(): RGBImageUncompressed {
        const c = new RGBImageUncompressed();
        c.init(this.xSize, this.ySize);
        c.data!.set(this.data!);
        return c;
    }
    public cloneToRgba(): RGBAImageUncompressed {
        const copy = new RGBAImageUncompressed();
        copy.init(this.xSize, this.ySize);
        const target = new RGBAPixel();
        for (let x = 0; x < this.xSize; x++)
            for (let y = 0; y < this.ySize; y++) {
                const source = this.getPixel(x, y);
                target.r = source.r;
                target.g = source.g;
                target.b = source.b;
                copy.putPixel(x, y, target);
            }
        return copy;
    }
    /** Java's current direct-buffer disposal implementation is intentionally a no-op. */
    public dispose(): void {}
}
